### PR TITLE
refactor(tabset): new directives to allow more customization of tabset

### PR DESCRIPTION
**Pull request by divdavem**
I am opening this PR to start discussing about a way to refactor the tabset, to allow more customization of the markup.

This could help solving the following issues: #2461, #2264, #2186, #1567, #1436 and #465

A similar work could be done to refactor the accordion, to help solving similar issues on the accordion.

The idea is to let the developer write a markup very similar to the raw bootstrap markup, but with directives on each element so that it is simpler to write than the raw markup while still allowing better customization than the current `<ngb-tabset>`. As all elements are explicit, the markup can be fully customized by adding any other directive, CSS class or element anywhere if needed.

```html
<ul ngbTabset selfControlled #myTabset="ngbTabset">
  <li ngbTab>
    <a ngbTabLink>Tab 1 title</a>
    <ng-template ngbTabContent>Tab 1 content</ng-template>
  </li>
  <li ngbTab>
    <a ngbTabLink>Tab 2 title</a>
    <ng-template ngbTabContent>Tab 2 content</ng-template>
  </li>
</ul>
<div [ngbTabsetOutlet]="myTabset"></div>
```

In this PR, I have implemented those new directives and I have also refactored the current tabset to use them.

The `selfControlled` directive manages the currently selected tab and makes sure that one valid tab is always selected.
Alternatively, it is possible to bind `[(activeId)]`, thus giving a solution for #2461.
It would also be possible to reimplement the `routerControlled` directive of #2603 to be compatible with this set of directives.

I have not added new tests yet but existing tests are passing with the refactored code.

Of course, I am not expecting this PR to be merged right away, but I am opening it more to start a discussion. Feel free to comment and share your ideas!

Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [ ] added/updated any applicable tests.
 - [ ] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@divdavem thnx for the PR - I very much like the direction and IMO we should be structuring things in the way that it is easy to do simple / common things and _possible_ to do complex / advanced things.

I haven't looked at the code in the details yet but the things that surprised me while doing first review is the presence of the `selfControlled` directive. What is the reasoning behind adding it and why it is not part of the `ngbTabset`? Is it to support routing?

File: src/tabset/tabset-directives.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  host: {
+    'role': 'tablist',
+    '[class]':
+        'className + \' nav nav-\' + type + (orientation == \'horizontal\'?  \' \' + justifyClass : \' flex-column\')'
+  }
+})
+export class NgbTabsetDirective {
+  justifyClass: string;
+
+  @Input('class') className = '';
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply src/tabset/tabset-directives.ts]: Same for `justifyClass`

**Comment by pkozlowski-opensource** [comment]
@divdavem I think I like it more an more :-) This solves so many of the existing issues! We should work towards merging it and I would like to propose that we hold off merging till we've got support for animations (this will impact the existing directives).

I would also very much like to deprecate the existing `<ngb-tabset>` directive and combine your ideas with the existing ideas so we can mix and match them in a single instance. Something along those lines:

```html
<ul ngbTabset selfControlled #myTabset="ngbTabset">
  <li ngbTab> <!-- I want to have full control here -->
    <a ngbTabLink>Tab 1 title</a>
    <ng-template ngbTabContent>Tab 1 content</ng-template>
  </li>
  <ngb-tab title="Simple tab with defaults">
    <ng-template ngbTabContent>Tab 2 content</ng-template>
  </ngb-tab>
</ul>
<div [ngbTabsetOutlet]="myTabset"></div>
```

Not sure if this can work, but would be interesting to explore.

Once again, thnx for this proposal!

**Comment by benouat** [comment]
I'd like to ping people on that one... cc @maxokorokov @pkozlowski-opensource @divdavem 

Could we proceed to move on ? like rebasing/preparing for landing in next feature release maybe ?
WDYT ?

**Comment by benouat** [comment]
As myself requested a month ago, I'll go ahead and decide to proceed with next step! 🤣 

@divdavem could you have a look at the 2 'left-without-response' comments from @pkozlowski-opensource and proceed with a rebase please ? thanks!

**Comment by lbod** [comment]
Hi.. I see this is on master now, sorry to chase but I couldn't see a roadplan for when you expect to release 4.2.0

**Comment by tomscholz** [comment]
Hey folks :)
What needs to be done to get this shipped? How can help?

- Cheers, Tom

**Comment by maxokorokov** [comment]
I cherry-picked this as a base for #1584, so closing.

P.S. Oops and sorry for pushing into your remote accidentally :/

**Pull request closed without merging**